signer/threshold_validator.go

```diff
@@ -103,12 +103,90 @@ func (pv *ThresholdValidator) mustLoadChainState(chainID string) ChainSignState
 // sign process if it is greater than the current high watermark. A mutex is used to avoid concurrent
 // state updates. The disk write is scheduled in a separate goroutine which will perform an atomic write.
 // pendingDiskWG is used upon termination in pendingDiskWG to ensure all writes have completed.
-func (pv *ThresholdValidator) SaveLastSignedStateInitiated(chainID string, signState SignStateConsensus) error {
+func (pv *ThresholdValidator) SaveLastSignedStateInitiated(chainID string, block *Block) ([]byte, time.Time, error) {
 	css := pv.mustLoadChainState(chainID)
 
-	css.lastSignStateInitiatedMutex.Lock()
-	defer css.lastSignStateInitiatedMutex.Unlock()
-	return css.lastSignStateInitiated.Save(signState, &pv.pendingDiskWG)
+	height, round, step := block.Height, block.Round, block.Step
+
+	err := css.lastSignStateInitiated.Save(NewSignStateConsensus(height, round, step), &pv.pendingDiskWG)
+	if err == nil {
+		// good to sign
+		return nil, time.Time{}, nil
+	}
+
+	existingSignature, existingTimestamp, sameBlockErr := pv.getExistingBlockSignature(chainID, block)
+
+	if _, ok := err.(*SameHRSError); ok {
+		// Wait for last sign state signature to be the same block
```

Comment:
I don't think this comment fits here; I think it's the overall logic of this block, but it looks like it details the double-error check adjacent to the comment.

Separate with a blank line, or add a new comment about why the errors are checked in this particular way?